go-controller/pkg/clustermanager/userdefinednetwork/controller.go

```diff
@@ -112,19 +118,109 @@ func (c *Controller) reconcile(key string) error {
 
-	syncErr := c.syncUserDefinedNetwork(udnCopy, nadCopy)
+	nadCopy, syncErr := c.syncUserDefinedNetwork(udnCopy, nadCopy)
 
 	updateStatusErr := c.updateUserDefinedNetworkStatus(udnCopy, nadCopy, syncErr)
 
 	return errors.Join(syncErr, updateStatusErr)
 }
 
-func (c *Controller) syncUserDefinedNetwork(udn *userdefinednetworkv1.UserDefinedNetwork, nad *netv1.NetworkAttachmentDefinition) error {
-	//TODO: implement sync UDN objects logic
-	return errors.New("implement me")
+func (c *Controller) syncUserDefinedNetwork(udn *userdefinednetworkv1.UserDefinedNetwork, nad *netv1.NetworkAttachmentDefinition) (*netv1.NetworkAttachmentDefinition, error) {
+	if udn == nil {
+		return nil, nil
+	}
+
+	desiredNAD, err := c.renderNadFn(udn)
+	if err != nil {
+		return nil, fmt.Errorf("failed to generate NetworkAttachmentDefinition: %w", err)
+	}
+	if nad == nil {
+		nad, err = c.nadClient.K8sCniCncfIoV1().NetworkAttachmentDefinitions(udn.Namespace).Create(context.Background(), desiredNAD, metav1.CreateOptions{})
+		if err != nil {
+			return nil, fmt.Errorf("failed to create NetworkAttachmentDefinition: %w", err)
+		}
+		klog.Infof("Created NetworkAttachmentDefinition [%s/%s]", nad.Namespace, nad.Name)
+		return nad, nil
+	}
+
+	if !metav1.IsControlledBy(nad, udn) {
+		return nil, fmt.Errorf("foreign NetworkAttachmentDefinition with the desired name already exist [%s/%s]", nad.Namespace, nad.Name)
+	}
+
+	if !reflect.DeepEqual(nad.Spec, desiredNAD.Spec) ||
+		!reflect.DeepEqual(nad.ObjectMeta.OwnerReferences, desiredNAD.ObjectMeta.OwnerReferences) {
+		nad.Spec.Config = desiredNAD.Spec.Config
+		nad.ObjectMeta.OwnerReferences = desiredNAD.ObjectMeta.OwnerReferences
+		nad, err = c.nadClient.K8sCniCncfIoV1().NetworkAttachmentDefinitions(nad.Namespace).Update(context.Background(), nad, metav1.UpdateOptions{})
+		if err != nil {
+			return nil, fmt.Errorf("failed to update NetworkAttachmentDefinition: %w", err)
+		}
+		klog.Infof("Updated NetworkAttachmentDefinition [%s/%s]", nad.Namespace, nad.Name)
+	}
+
+	return nad, nil
 }
 
 func (c *Controller) updateUserDefinedNetworkStatus(udn *userdefinednetworkv1.UserDefinedNetwork, nad *netv1.NetworkAttachmentDefinition, syncError error) error {
-	//TODO: implement UDN status update logic
-	return errors.New("implement me")
+	if udn == nil {
+		return nil
+	}
+
+	networkReadyCondition := newNetworkReadyCondition(nad, syncError)
+
+	conditions, updated := updateCondition(udn.Status.Conditions, networkReadyCondition)
+
+	if updated {
+		var err error
+		udnApplyConf := udnapplyconfkv1.UserDefinedNetwork(udn.Name, udn.Namespace).
+			WithStatus(udnapplyconfkv1.UserDefinedNetworkStatus().
+				WithConditions(conditions...))
+		opts := metav1.ApplyOptions{FieldManager: "user-defined-network-controller"}
+		udn, err = c.udnClient.K8sV1().UserDefinedNetworks(udn.Namespace).ApplyStatus(context.Background(), udnApplyConf, opts)
+		if err != nil {
+			return fmt.Errorf("failed to update UserDefinedNetwork status: %w", err)
+		}
```

Comment:
suggestion to avoid errors on deleted UDN update
```
if err != nil {
			if apierrors.IsNotFound(err) {
				return nil
			}
			return fmt.Errorf("failed to update UserDefinedNetwork status: %w", err)
		}
```

Reply:
Done

Reply:
your change is not exactly the same, as it will now log the following line
```
klog.Infof("Updated status UserDefinedNetwork [%s/%s]", udn.Namespace, udn.Name)
```
while in reality status was not updated and the UDN is deleted.